python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py

```diff
@@ -75,62 +76,58 @@ def processAlgorithm(self, progress):
-                        feature_list = self.extractAsMulti(inGeom)
-                        multi_feature.extend(feature_list)
-
-                    current += 1
-                    progress.setPercentage(int(current * total))
-
-                outFeat.setAttributes(attrs)
-                outGeom = QgsGeometry(self.convertGeometry(multi_feature,
-                                                           vType))
-                outFeat.setGeometry(outGeom)
-                writer.addFeature(outFeat)
-
-            del writer
-        else:
-            raise GeoAlgorithmExecutionException(
-                self.tr('At least two features must have same attribute '
-                        'value! Please choose another field...'))
-
-    def singleToMultiGeom(self, wkbType):
-        try:
-            if wkbType in (QgsWkbTypes.Point, QgsWkbTypes.MultiPoint,
-                           QgsWkbTypes.Point25D, QgsWkbTypes.MultiPoint25D):
-                return QgsWkbTypes.MultiPoint
-            elif wkbType in (QgsWkbTypes.LineString, QgsWkbTypes.MultiLineString,
-                             QgsWkbTypes.MultiLineString25D,
-                             QgsWkbTypes.LineString25D):
-
-                return QgsWkbTypes.MultiLineString
-            elif wkbType in (QgsWkbTypes.Polygon, QgsWkbTypes.MultiPolygon,
-                             QgsWkbTypes.MultiPolygon25D, QgsWkbTypes.Polygon25D):
-
-                return QgsWkbTypes.MultiPolygon
+
+        for i in unique:
+            if fieldType in [QVariant.Int, QVariant.Double, QVariant.UInt, QVariant.LongLong, QVariant.ULongLong]:
+                expr = '"%s" = %s' % (fieldName, i)
+            elif fieldType == QVariant.String:
+                expr = """"%s" like '%s'""" % (fieldName, i)
+            elif fieldType in [QVariant.Date, QVariant.DateTime]:
+                expr = """"%s" = '%s'""" % (fieldName, i)
+            else:
+                raise GeoAlgorithmExecutionException(
+                    self.tr('Unsupported field type "%s"' % fields[idx].typeName()))
+
+            qExp = QgsExpression(expr)
+            if not qExp.hasParserError():
+                qReq = QgsFeatureRequest(qExp).setSubsetOfAttributes([])
             else:
-                return QgsWkbTypes.Unknown
-        except Exception:
-            pass
+                raise GeoAlgorithmExecutionException(qExp.parserErrorString())
+
+            multi_feature = []
+            first = True
+            features = vector.features(layer, qReq)
+            for inFeat in features:
+                atMap = inFeat.attributes()
+                idVar = atMap[index]
+                if str(idVar).strip() == str(i).strip():
+                    if first:
+                        attrs = atMap
+                        first = False
+                    inGeom = inFeat.geometry()
+                    vType = inGeom.type()
+                    feature_list = self.extractAsMulti(inGeom)
```

Comment:
This part still breaks z/m/curves. I think it should use QgsGeometry::collectGeometry instead